File: src/kalderstam/neural/functions/activation_functions.py
#A few activation functions
import math
from math import exp

class logsig():
    def function(self, x):
        return 1 / (1 + exp(-x))
    def derivative(self, x):
        #return exp(x)/((exp(x) + 1)**2)
        y = self.function(x)
        return y*(1 - y)

class linear():
    def function(self, x):
        return x
    def derivative(self, x):
        return 1

class tanh():
    def function(self, x):
        return math.tanh(x)
    def derivative(self, x):
        #return (1/math.cosh(x))**2 #sech(x)^2
        y = self.function(x)
        return (1 - y)*(1 + y)

File: src/C_ext/activation_functions.c
#include "activation_functions.h"

double linear(double input)
{
	return input;
}

double linear_derivative(double input)
{
	return 1.0;
}

double logsig(double input)
{
	return 1 / (1 + exp(-input));
}

double logsig_derivative(double input)
{
	double y = logsig(input);
	return y * (1 - y);
}

//tanh defined in math.h which is included, do not override it

double tanh_derivative(double input)
{
	double y = tanh(input);
	return (1 - y) * (1 + y);
}
